File: lct-frontend/assets/ts/Modules/Models/Level.ts
import {ProgressContainer} from "./ProgressContainer.ts";

export class Level {
    id: number;
    name: string;
    author: string;
    attempts: number;
    priority: number;
    mainProgress: number;
    difficulty: string;
    status: Statuses;
    progressContainers: ProgressContainer[];
    constructor(name: string, author: string,
                difficulty: string, status:Statuses = Statuses.Active) {
        this.id = 0;
        this.name = name;
        this.author = author;
        this.attempts = 0;
        this.difficulty = difficulty;
        this.status = status;
        this.mainProgress = 0;
        this.priority = 0;

        this.progressContainers = [];
    }
}

export enum Difficulties {
    Easy = 'easy',
    Normal = 'normal',
    Hard = 'hard',
    Harder = 'harder',
    Insane = 'insane',
    EasyDemon = 'easy-demon',
    MediumDemon = 'medium-demon',
    HardDemon = 'hard-demon',
    InsaneDemon = 'insane-demon',
    ExtremeDemon = 'extreme-demon',
    Default = 'na',
}

export enum Statuses {
    Active = 'active',
    Abandoned = 'abandoned',
    Completed = 'completed',
}

export function getStatusColor(status: string): string {
    switch (status) {
        case 'active':
            return 'green';
        case 'abandoned':
            return 'orange';
        default:
            return 'black';
    }    }

export function getStatusText(status: string): string {
    switch (status) {
        case 'active':
            return 'Active';
        case 'abandoned':
            return 'Abandoned';
        case 'completed':
            return 'Completed';
        default:
            return 'N/A';
    }
}

File: LevelsCompletionTracker.Core/Model/Level.cs
using System.Net.Http.Headers;
using LevelsCompletionTracker.Shared.Output;

namespace LevelsCompletionTracker.Core.Model
{
    public static class LevelStatus
    {
        public static string Active => "active";
        public static string Abandoned => "abandoned";
        public static string Completed => "completed";
    }

    public class Level
    {
        private int mainProgress = 0;
        private string status = "";

        public int Id { get; set; }
        public string Name { get; set; } = null!;
        public string Author { get; set; } = null!;
        public string Difficulty { get; set; } = null!;
        public int Attempts { get; set; }
        public int Priority { get; set; }

        public int MainProgress
        {
            get => mainProgress;
            set
            {
                if (value < 0 || value > 100)
                {
                    throw new ArgumentOutOfRangeException("", "Progress is out of range");
                }

                if (value == 100)
                {
                    Status = LevelStatus.Completed;
                }
                else
                {
                    Status = LevelStatus.Active;
                }

                mainProgress = value;
            }
        }

        public string Status
        {
            get => status;
            set
            {
                if (string.IsNullOrWhiteSpace(value) || (!string.Equals(value, LevelStatus.Active) && !string.Equals(value, LevelStatus.Abandoned) && !string.Equals(value, LevelStatus.Completed)))
                {
                    throw new ArgumentOutOfRangeException("", "Status has only 3 variants: active, abandoned and completed");
                }

                status = value;

            }
        }

        public ICollection<ProgressContainer> ProgressContainers { get; set; } = new List<ProgressContainer>();
    }
}
